File: frontend/src/lib/raCodes.ts
/**
 * GLEIF Registration Authority codes for the 21 countries that have
 * OpenCheck adapters. Used to scope the national-ID reverse lookup to
 * a single registry and avoid false matches from coincidental ID
 * collisions across registries.
 *
 * filter[entity.registeredAt]=<raCode> is the GLEIF API parameter.
 * Reference: https://www.gleif.org/en/about-lei/code-lists/gleif-registration-authorities-list
 *
 * Every code re-verified 2026-08-28 against live GLEIF records — reading
 * registeredAt.id off real entities, not off the RA catalogue, because the
 * catalogue only says which authorities exist, not which one GLEIF actually
 * files a country's companies under. Eleven of the twenty were wrong, and
 * each pointed at a real but different authority, so the country picker
 * scoped the reverse lookup to a registry the company was not registered at
 * and returned no match. Norway pointed at India's MCA (RA000394) and Sweden
 * at Singapore's ACRA (RA000523); both had already been corrected in the
 * backend and in CLAUDE.md without this file being updated.
 *
 * Keep in step with _RA_BY_COUNTRY in backend/opencheck/routers/lookup.py —
 * backend/tests/test_ra_codes.py parses THIS FILE and fails if they diverge.
 */

export interface RaEntry {
  raCode: string;
  countryName: string;
  /** Short label for the input field, e.g. "Companies House number" */
  idLabel: string;
  /** Placeholder value shown in the text input */
  placeholder: string;
  /** One-line format hint shown below the input */
  formatHint: string;
  /**
   * Optional regex used for client-side format validation.
   * Applied to the trimmed input value. Absence means "no strict check"
   * (used for countries with variable-length or complex ID formats).
   * Validation is advisory — a mismatch shows a warning but never blocks
   * submission, since GLEIF may store the ID in a normalised form.
   */
  formatPattern?: RegExp;
  /**
   * Countries whose companies are filed under more than one GLEIF authority,
   * where the registration number's prefix says which. Rules are tried in
   * order; a number matching none of them keeps `raCode`.
   *
   * Companies House is the only one OpenCheck covers. This matters more than
   * it looks: the GB format hint below explicitly invites an `SC` or `NI`
   * number, and until Phase 141 every one of them was scoped to the England &
   * Wales authority, which is the one registry those companies are guaranteed
   * *not* to be in. The query returned nothing, and nothing is what an absent
   * company also returns.
   *
   * Mirrors SUB_REGISTRIES in backend/opencheck/ra_codes.py —
   * backend/tests/test_ra_codes.py parses this file and fails if they diverge.
   */
  subRegistries?: { prefixes: string[]; raCode: string; label: string }[];
}

export const RA_CODES: Record<string, RaEntry> = {
  GB: {
    raCode: "RA000585",
    countryName: "United Kingdom",
    idLabel: "Companies House number",
    placeholder: "02000048",
    formatHint: "8 characters — digits or two-letter prefix (OC, SC, NI…) + 6 digits",
    // 8 pure digits OR two uppercase letters + 6 digits (total 8 chars).
    formatPattern: /^(?:\d{8}|[A-Z]{2}\d{6})$/i,
    subRegistries: [
      // Scottish limited companies are SC; Scottish limited partnerships and
      // qualifying partnerships are SO and SF.
      { prefixes: ["SC", "SO", "SF"], raCode: "RA000587", label: "Companies House — Scotland" },
      // Northern Irish companies are NI; NC and R0 are older registrations
      // carried over from the pre-2009 Belfast registry.
      { prefixes: ["NI", "NC", "R0"], raCode: "RA000586", label: "Companies House — Northern Ireland" },
    ],
  },
  NL: {
    raCode: "RA000463",
    countryName: "Netherlands",
    idLabel: "KvK number",
    placeholder: "34362985",
    formatHint: "8 digits",
    formatPattern: /^\d{8}$/,
  },
  NO: {
    raCode: "RA000472",
    countryName: "Norway",
    idLabel: "Organisation number (orgnr)",
    placeholder: "923609016",
    formatHint: "9 digits",
    formatPattern: /^\d{9}$/,
  },
  NZ: {
    raCode: "RA000466",
    countryName: "New Zealand",
    idLabel: "Company number",
    placeholder: "1166320",
    formatHint: "Companies Register number (digits)",
    formatPattern: /^\d{1,9}$/,
  },
  DK: {
    raCode: "RA000170",
    countryName: "Denmark",
    idLabel: "CVR number",
    placeholder: "36213728",
    formatHint: "8 digits",
    formatPattern: /^\d{8}$/,
  },
  SE: {
    raCode: "RA000544",
    countryName: "Sweden",
    idLabel: "Organisation number",
    placeholder: "5560985801",
    formatHint: "10 digits, optionally written as NNNNNN-NNNN",
    // 10 pure digits OR NNNNNN-NNNN (with dash).
    formatPattern: /^\d{10}$|^\d{6}-\d{4}$/,
  },
  FR: {
    raCode: "RA000189",
    countryName: "France",
    idLabel: "SIREN number",
    placeholder: "542107651",
    formatHint: "9 digits",
    formatPattern: /^\d{9}$/,
  },
  BE: {
    raCode: "RA000025",
    countryName: "Belgium",
    idLabel: "CBE / KBO number",
    placeholder: "0403838524",
    formatHint: "10 digits",
    formatPattern: /^\d{10}$/,
  },
  BR: {
    raCode: "RA000681",
    countryName: "Brazil",
    idLabel: "CNPJ",
    placeholder: "33.000.167/0001-01",
    formatHint: "14 digits (CNPJ), with or without punctuation",
    // 14 plain digits OR the punctuated XX.XXX.XXX/XXXX-XX form.
    formatPattern: /^\d{14}$|^\d{2}\.\d{3}\.\d{3}\/\d{4}-\d{2}$/,
  },
  IE: {
    raCode: "RA000402",
    countryName: "Ireland",
    idLabel: "CRO number",
    placeholder: "012345",
    formatHint: "Up to 6 digits",
    formatPattern: /^\d{1,6}$/,
  },
  PL: {
    raCode: "RA000484",
    countryName: "Poland",
    idLabel: "KRS number",
    placeholder: "0000037171",
    formatHint: "10 digits",
    formatPattern: /^\d{10}$/,
  },
  AT: {
    raCode: "RA000017",
    countryName: "Austria",
    idLabel: "Firmenbuchnummer",
    placeholder: "FN123456a",
    formatHint: "FN + digits + letter suffix (e.g. FN 237338 p or FN123456a)",
    // Optional "FN" prefix + optional space + digits + optional space + single letter suffix.
    formatPattern: /^(?:FN\s*)?\d+\s*[a-z]?$/i,
  },
  EE: {
    raCode: "RA000181",
    countryName: "Estonia",
    idLabel: "Registration code",
    placeholder: "10138896",
    formatHint: "8 digits",
    formatPattern: /^\d{8}$/,
  },
  LV: {
    raCode: "RA000423",
    countryName: "Latvia",
    idLabel: "Registration number",
    placeholder: "40003571815",
    formatHint: "11 digits",
    formatPattern: /^\d{11}$/,
  },
  LT: {
    raCode: "RA000430",
    countryName: "Lithuania",
    idLabel: "JAR code",
    placeholder: "302511363",
    formatHint: "9 digits",
    formatPattern: /^\d{9}$/,
  },
  MT: {
    raCode: "RA000443",
    countryName: "Malta",
    idLabel: "Registration number (C-number)",
    placeholder: "C 113927",
    formatHint: "Letter prefix + number, e.g. C 12345",
    // Short letter prefix (commonly C) + digits, with or without a space.
    formatPattern: /^[A-Z]{1,3}\s*\d+$/i,
  },
  SK: {
    raCode: "RA000526",
    countryName: "Slovakia",
    idLabel: "IČO number",
    placeholder: "31320155",
    formatHint: "8 digits",
    formatPattern: /^\d{8}$/,
  },
  HR: {
    raCode: "RA000156",
    countryName: "Croatia",
    idLabel: "OIB",
    placeholder: "30420566661",
    formatHint: "11 digits",
    formatPattern: /^\d{11}$/,
  },
  SG: {
    raCode: "RA000523",
    countryName: "Singapore",
    idLabel: "UEN",
    placeholder: "196700240H",
    formatHint: "9–10 alphanumeric characters",
    // Local companies: 9 digits + check letter. Foreign/other: various.
    // Accept 9-10 alphanumeric chars as the common denominator.
    formatPattern: /^[A-Z0-9]{9,10}$/i,
  },
  CA: {
    raCode: "RA000072",
    countryName: "Canada",
    idLabel: "Corporation number",
    placeholder: "1234567",
    formatHint: "7–9 digits (federal corporations)",
    formatPattern: /^\d{7,9}$/,
  },
  GR: {
    raCode: "RA000685",
    countryName: "Greece",
    idLabel: "Αριθμός ΓΕΜΗ (GEMI number)",
    placeholder: "160228803000",
    // ΓΕΜΗ numbers are 9-12 digits. GLEIF stores some zero-padded to 12 and
    // some not, and the ΓΕΜΗ API accepts either, so do not normalise here.
    formatHint: "9–12 digits",
    formatPattern: /^\d{9,12}$/,
  },
};

/**
 * Alphabetical by country name. UK is the default selected value (set in
 * App.tsx state) but sits in its natural A–Z position in the list.
 *
 * Derived from RA_CODES rather than hand-listed: the hand-listed version had
 * silently omitted New Zealand, so a country with a correct RA code, a format
 * hint and a working backend mapping could not be picked at all. Greece was
 * missing from the file entirely, having been added to the backend map when
 * the ΓΕΜΗ adapter shipped. Neither produced an error — an option that does
 * not exist simply is not there to notice.
 */
export const COUNTRY_OPTIONS: { code: string; entry: RaEntry }[] = Object.entries(RA_CODES)
  .map(([code, entry]) => ({ code, entry }))
  .sort((a, b) => a.entry.countryName.localeCompare(b.entry.countryName, "en"));

/**
 * Returns true if `value` is empty, the country has no pattern defined,
 * or `value` matches the country's formatPattern.
 *
 * Validation is advisory — callers should warn but not block submission.
 */
export function validateNationalId(countryCode: string, value: string): boolean {
  const trimmed = value.trim();
  if (!trimmed) return true;
  const pattern = RA_CODES[countryCode]?.formatPattern;
  if (!pattern) return true;
  return pattern.test(trimmed);
}

/**
 * The RA code a reverse lookup should be scoped to — the country's authority,
 * refined by the registration number where the country has sub-registries.
 *
 * Callers should always pass the number they have. `RA_CODES[c].raCode` alone
 * answers "which registry is this country's", which is not the same question
 * as "which registry is this company in", and for GB the two differ for every
 * Scottish and Northern Irish company.
 *
 * Returns "" for an unknown country, which leaves the GLEIF query unscoped —
 * a wider search rather than a wrong one.
 */
export function raCodeFor(countryCode: string, value = ""): string {
  const entry = RA_CODES[countryCode];
  if (!entry) return "";
  const trimmed = value.trim().toUpperCase();
  if (trimmed) {
    for (const rule of entry.subRegistries ?? []) {
      if (rule.prefixes.some((p) => trimmed.startsWith(p))) return rule.raCode;
    }
  }
  return entry.raCode;
}

File: backend/opencheck/ra_codes.py
"""Country → GLEIF Registration Authority dispatch, in one place.

Every reverse lookup OpenCheck runs — national registration number to LEI —
has to answer one question first: *which* registration authority should the
GLEIF query be scoped to? Getting it wrong does not raise; it appends
``filter[entity.registeredAt]=<wrong code>`` and returns an empty result set,
which reads as an absent company rather than a defect.

That failure mode is why this module exists. The mapping had been written out
four times — ``routers/lookup.py``, ``routers/search.py``, ``sources/gleif.py``
and ``frontend/src/lib/raCodes.ts`` — and by 2026-08-28 every copy had drifted
separately: twenty wrong values between them, each one a real code belonging to
a different authority. Phase 140 corrected all four. Phase 141 removes the
premise, because correcting copies does not stop them diverging again.

**Two questions, not one.** Phase 140 fixed the values and left the shape,
and the shape was the remaining half of the bug: the country map and the
prefix rule lived in different modules and only one entry point consulted
both. ``/search``'s Companies House → LEI bridge applied the prefix rule;
``/resolve-national-id`` went straight to the flat country map, so
``country="GB"`` sent a Scottish number to the England & Wales authority and
found nothing. Both now go through :func:`ra_code_for`.

**Sub-registries are data, not a special case.** GB is the only country that
needs one today, but expressing it as a table rather than an ``if`` means the
frontend can mirror the same declaration and ``tests/test_ra_codes.py`` can
check the two agree — which is the only mechanism that has ever caught these
copies drifting.

Every code here was verified against **live GLEIF records** — reading
``registeredAt.id`` off real entities via ``filter[entity.legalAddress.country]``
— and not against the RA catalogue, which only says which authorities exist.
``filter[country]`` on that catalogue endpoint is silently ignored and returns
the unfiltered global list, so spot-checking a code that way appears to confirm
whatever you already believed. That is the most likely origin of the original
values, and it is worth knowing before changing anything below.
"""

from __future__ import annotations

from typing import NamedTuple

__all__ = [
    "RA_BY_COUNTRY",
    "SUB_REGISTRIES",
    "SubRegistry",
    "ch_ra_code",
    "ra_code_for",
]


#: Country (ISO 3166-1 alpha-2) → the GLEIF Registration Authority code that
#: scopes a reverse lookup for that country. Where a country files companies
#: under several authorities this is the **dominant** one; :data:`SUB_REGISTRIES`
#: refines it from the registration number. Pass an explicit ``ra_code`` to
#: target a specific authority directly.
RA_BY_COUNTRY: dict[str, str] = {
    "GB": "RA000585",  # UK Companies House (England & Wales) — see SUB_REGISTRIES
    "NL": "RA000463",  # KvK (Netherlands)
    "NO": "RA000472",  # Brønnøysund / Brreg (Norway)
    "NZ": "RA000466",  # Companies Office (New Zealand)
    "IE": "RA000402",  # CRO (Ireland)
    "LV": "RA000423",  # UR (Latvia)
    "LT": "RA000430",  # JAR (Lithuania)
    "FR": "RA000189",  # Sirene / INSEE (France)
    "SE": "RA000544",  # Bolagsverket (Sweden)
    "EE": "RA000181",  # ariregister (Estonia)
    "BE": "RA000025",  # BCE/KBO (Belgium)
    "AT": "RA000017",  # Firmenbuch (Austria)
    "PL": "RA000484",  # KRS (Poland)
    "SK": "RA000526",  # RPO (Slovakia)
    "SG": "RA000523",  # ACRA (Singapore)
    "CA": "RA000072",  # Corporations Canada
    "DK": "RA000170",  # CVR (Denmark)
    "HR": "RA000156",  # Sudski registar (Croatia)
    "MT": "RA000443",  # Malta Business Registry
    "BR": "RA000681",  # Receita Federal CNPJ (Brazil)
    "GR": "RA000685",  # ΓΕΜΗ — General Commercial Registry (Greece)
}


class SubRegistry(NamedTuple):
    """One authority within a country, selected by registration-number prefix."""

    prefixes: tuple[str, ...]
    ra_code: str
    label: str


#: Countries whose companies are filed under more than one authority, where the
#: registration number's prefix says which. Rules are tried in order; a number
#: matching none of them keeps the country's default from :data:`RA_BY_COUNTRY`.
#:
#: Companies House is the only such country OpenCheck covers. Its three
#: authorities are ``RA000585`` England & Wales, ``RA000586`` Northern Ireland
#: and ``RA000587`` Scotland — confirmed against real records (THON MARITIME
#: LTD, ``registeredAs "SC651281"``, sits under RA000587). ``RA000591`` is
#: **The Pensions Regulator** and not a company registry at all; it sat here as
#: Northern Ireland's code until Phase 140, and ``tests/test_ra_codes.py``
#: now fails if it reappears in any mapping.
SUB_REGISTRIES: dict[str, tuple[SubRegistry, ...]] = {
    "GB": (
        # Scottish limited companies are SC; Scottish limited partnerships and
        # qualifying partnerships are SO and SF.
        SubRegistry(("SC", "SO", "SF"), "RA000587", "Companies House — Scotland"),
        # Northern Irish companies are NI; NC and R0 are older registrations
        # carried over from the pre-2009 Belfast registry.
        SubRegistry(("NI", "NC", "R0"), "RA000586", "Companies House — Northern Ireland"),
    ),
}


def ch_ra_code(company_number: str) -> str:
    """Return the GLEIF Registration Authority code for a Companies House number.

    A thin, well-named alias for ``ra_code_for("GB", company_number)``, kept
    because the Companies House → LEI bridge in ``routers/search.py`` reads
    better with it and because ``opencheck.app`` re-exports the name.
    """
    return ra_code_for("GB", company_number)


def ra_code_for(country: str, number: str = "") -> str:
    """Return the RA code a reverse lookup should be scoped to.

    ``country`` is an ISO 3166-1 alpha-2 code; ``number`` is the national
    registration number, used only to pick between a country's sub-registries.
    Returns ``""`` for a country OpenCheck has no mapping for — the caller
    then queries GLEIF unscoped, which is a wider search rather than a wrong
    one.

    Passing no ``number`` yields the country's dominant authority. That is the
    right answer for "which registry is this country's", and the wrong one for
    "which registry is *this company* in" — so callers that have the number
    should always pass it.
    """
    code = country.strip().upper()
    default = RA_BY_COUNTRY.get(code, "")

    upper = (number or "").strip().upper()
    if upper:
        for rule in SUB_REGISTRIES.get(code, ()):
            if upper.startswith(rule.prefixes):
                return rule.ra_code

    return default
